### PR TITLE
move preprocess_logits_for_metrics before _nested_gather in trainer.e…

### LINKED ISSUE #22602: Preprocess/transform logits before gathering them for computing metrics.
### Feature request

In `trainer.evaluation_loop`, `preprocess_logits_for_metrics` should be executed before `_nested_gather` to avoid GPU OOM.

### Motivation

`preprocess_logits_for_metrics` processes logits after gathering them when do distributed learning. When training with large batch_size, token_length or vocab_size, gathering all logits to 1 node will cause to out of GPU memory. This preprocessing should be executed before `_nested_gather`.

### Your contribution

The main modification would be this in `trainer.evaluation_loop`:
```
           if logits is not None:
                logits = self._pad_across_processes(logits)
                if self.preprocess_logits_for_metrics is not None:
                    logits = self.preprocess_logits_for_metrics(logits, labels)
                logits = self._nested_gather(logits)
                preds_host = logits if preds_host is None else nested_concat(preds_host, logits, padding_index=-100)
```

### PR DESCRIPTION
**Pull request by ChenyangLiu**
# What does this PR do?

Fixes #22602 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
This is incorrect. The labels have already been gathered at line 3183 across devices, so the logits need to be gathered as well to match the labels.

**Comment by ChenyangLiu** [comment]
Sorry. In my training loop, the `preprocess_logits_for_metrics` do not use `labels`. I ignore the `labels` is gathered before. In the new commit, the code is
```
# Update containers on host
if loss is not None:
    losses = self._nested_gather(loss.repeat(batch_size))
    losses_host = losses if losses_host is None else torch.cat((losses_host, losses), dim=0)
if labels is not None:
    labels = self._pad_across_processes(labels)        
if inputs_decode is not None:
    inputs_decode = self._pad_across_processes(inputs_decode)
    inputs_decode = self._nested_gather(inputs_decode)
    inputs_host = (
        inputs_decode
        if inputs_host is None
        else nested_concat(inputs_host, inputs_decode, padding_index=-100)
    )
if logits is not None:
    logits = self._pad_across_processes(logits)
if self.preprocess_logits_for_metrics is not None and logits is not None:
    logits = self.preprocess_logits_for_metrics(logits, labels)
if labels is not None:
    labels = self._nested_gather(labels)
    labels_host = labels if labels_host is None else nested_concat(labels_host, labels, padding_index=-100)
if logits is not None:
    logits = self._nested_gather(logits)
    preds_host = logits if preds_host is None else nested_concat(preds_host, logits, padding_index=-100)
```
labels and logits should be padded first, then be preprocessed before gathering. In my use case, I trained BLOOM with 32 batch size, the gathered logits size is (32, 1024, 250000+), which takes 15G+ gpu memory and cause to OOM during evaluating.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3192,9 +3190,13 @@ def evaluation_loop(
                 )
             if logits is not None:
                 logits = self._pad_across_processes(logits)
+            if self.preprocess_logits_for_metrics is not None and logits is not None:
+                logits = self.preprocess_logits_for_metrics(logits, labels)
```
Comment (sgugger) [inline src/transformers/trainer.py:3194]:
```suggestion
                if self.preprocess_logits_for_metrics is not None:
                    logits = self.preprocess_logits_for_metrics(logits, labels)
```

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3192,9 +3190,13 @@ def evaluation_loop(
...
+            if labels is not None:
+                labels = self._nested_gather(labels)
+                labels_host = labels if labels_host is None else nested_concat(labels_host, labels, padding_index=-100)
+            if logits is not None:
```
Comment (sgugger) [inline src/transformers/trainer.py:3198]: Move the labels test after logits, so that the nested_gather line can be in the same `if logits is not None` as above.

**Review by sgugger — COMMENTED** [review]
Makes more sense like this, thanks for iterating! Left a couple more comments.

**Comment by ChenyangLiu** [comment]
@sgugger Done. Rewrite the code with your suggestion. Thanks.

File: src/transformers/trainer.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/trainer.py:3198]: This is still pending. You can move line 3199 just after line 3194 and remove this new `if logits is not None` entirely.

Comment (ChenyangLiu) [reply src/transformers/trainer.py]: fix

**Review by sgugger — APPROVED** [review]
Thanks a lot!
